### PR TITLE
Implement delete and versionless update

### PR DESCRIPTION
**Pull request by adleong**
* Implement delete
* Don't send If-Match header on PUT

### CODE REVIEW

**Comment by adleong** [comment]
I don't understand that CI failure and can't reproduce it :(
It seems like it's using an out-of-date version of controller.go?

File: namer/controller.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ type (
 		Get(name string) (*VersionedDtab, error)
 		Create(name string, dtabstr string) (Version, error)
 		Delete(name string) error
-		Update(name string, dtabstr string, version Version) (Version, error)
+		Update(name string, dtabstr string) (Version, error)
```
Comment (olix0r) [inline namer/controller.go]: It seems like we'd still want update to take a version -- we just want another method Put that doesn't?  or we want to support nil versions?

We should think of this as a pure API client independent of the CLI.

File: cmd/dtab.go [hunk 2]
```diff
@@ -108,12 +108,11 @@ var (
 	}
 
 	dtabUpdateCmd = &cobra.Command{
-		Use:     "update [name] [file]",
+		Use:     "update [name] [file] [version]",
```
Comment (adleong) [inline cmd/dtab.go]: How can we indicate that version is optional?  I'm used to seeing square brackets for optional stuff, but everything is already in square brackets here...

Comment (siggy) [inline cmd/dtab.go]: maybe omit square brackets for required arguments?

```bash
$ man cp
CP(1)                     BSD General Commands Manual                    CP(1)

NAME
     cp -- copy files

SYNOPSIS
     cp [-R [-H | -L | -P]] [-fi | -n] [-apvX] source_file target_file
     cp [-R [-H | -L | -P]] [-fi | -n] [-apvX] source_file ... target_directory
```

File: cmd/dtab.go [hunk 3]
```diff
@@ -123,18 +122,20 @@ var (
...
 				}
-				_, err = ctl.Update(name, dtabstr, versioned.Version)
 				if err != nil {
 					return err
 				}
 				fmt.Printf("Updated %s\n", name)
 				return nil
-
-			default:
+			} else {
```
Comment (siggy) [inline cmd/dtab.go]: may be clearer with something like:
```go
version := ""
switch len(args) {
case 3:
	version = namer.Version(args[2])
	fallthrough
case 2:
	ctl, err := getController()
	if err != nil {
		return err
	}
	name := args[0]
	dtabstr, err := readDtabPath(args[1])
	if err != nil {
		return err
	}
	_, err = ctl.Update(name, dtabstr, version)
	if err != nil {
		return err
	}
	fmt.Printf("Updated %s\n", name)
	return nil
default:
	return errors.New("update requires a name and file path")
}
```

**Comment by siggy** [comment]
:+1: :star: confirmed works on new dcos cluster / 0.3.1. my comment was tioli.

File: cmd/dtab.go (hunk 3 above)
Comment (adleong) [inline cmd/dtab.go]: That's way better, thanks.  I think the part of my brain that knows about mutable variables has completely atrophied.

File: cmd/dtab.go [hunk 4]
```diff
@@ -108,34 +108,33 @@ var (
...
-		Use:     "update [name] [file]",
+		Use:     "update [name] [file] [version]",
 		Aliases: []string{"up"},
 		Short:   "Update a delegation table.",
 		RunE: func(cmd *cobra.Command, args []string) error {
+			version := namer.Version("")
 			switch len(args) {
+			case 3:
+			    version = namer.Version(args[2])
+			    fallthrough
```
Comment (klingerf) [inline cmd/dtab.go]: Oooh TIL

**Comment by klingerf** [comment]
:star: :tanabata_tree: Looks good

File: cmd/dtab.go (hunk 2 above)
Comment (olix0r) [inline cmd/dtab.go]: perhaps `update NAME FILE [version]`

```
:; kubectl apply --help
Apply a configuration to a resource by filename or stdin.
The resource will be created if it doesn't exist yet. 

JSON and YAML formats are accepted.

Usage:
  kubectl apply -f FILENAME [flags]
```

File: cmd/dtab.go [hunk 5]
```diff
@@ -108,34 +108,33 @@ var (
...
+			    }
+			    _, err = ctl.Update(name, dtabstr, version)
+			    if err != nil {
+			        return err
+			    }
+			    fmt.Printf("Updated %s\n", name)
+			    return nil
 			default:
-				return errors.New("update requires a name and file path")
+			    return errors.New("update requires a name and file path")
```
Comment (olix0r) [inline cmd/dtab.go]: Configure your editor to run gofmt on source code.  Tabs, not spaces.

File: cmd/dtab.go (hunk 2 above)
Comment (olix0r) [inline cmd/dtab.go]: Also, I have a slight preference for using flags for optional arguments.  e.g. `update -v VERSION NAME FILE` or `update NAME FILE`.  I could go either way in this case, though.

**Comment by gtcampbell** [comment]
:+1: (caveat: I am by no means a Go expert)

**Comment by gtcampbell** [comment]
@adleong: I think if you rebase against master, that will fix the pullapprove settings.

**Comment by olix0r** [comment]
:star:

**Comment by olix0r** [comment]
@esbie would you mind shippiting this?  hopefully this is the last namerctl review where you have sole approval

**Comment by esbie** [comment]
⭐

**Pull request merged**